app/routes/session.py

```diff
@@ -129,17 +134,39 @@ def login() -> Response:
 
     cookie_session["language_code"] = metadata.language_code
 
-    # Type ignore: survey_id and either ru_ref or qid are required for schemas that use supplementary data
-    if dataset_id := metadata["sds_dataset_id"]:
-        get_supplementary_data(
-            dataset_id=dataset_id,
-            unit_id=metadata["ru_ref"] or metadata["qid"],  # type: ignore
-            survey_id=metadata["survey_id"],  # type: ignore
-        )
-
     return redirect(url_for("questionnaire.get_questionnaire"))
 
 
+def _set_questionnaire_supplementary_data(
+    questionnaire_store: QuestionnaireStore, metadata: MetadataProxy
+) -> None:
+    """
+    If the survey metadata has an sds dataset id, and it either doesn't match what it stored, or there is no stored supplementary data
+    then fetch it and add it to the store
+    """
+    if not (new_sds_dataset_id := metadata["sds_dataset_id"]):
+        return
+
+    existing_sds_dataset_id = (
+        questionnaire_store.metadata.survey_metadata["sds_dataset_id"]
+        if questionnaire_store.metadata and questionnaire_store.metadata.survey_metadata
+        else None
+    )
+
+    if existing_sds_dataset_id == new_sds_dataset_id:
+        # no need to fetch again
+        return
+
+    supplementary_data = get_supplementary_data(
+        # Type ignore: survey_id and either ru_ref or qid are required for schemas that use supplementary data
+        dataset_id=new_sds_dataset_id,
+        unit_id=metadata["ru_ref"] or metadata["qid"],  # type: ignore
+        survey_id=metadata["survey_id"],  # type: ignore
+    )
+    logger.info("supplementary data fetched")
```

Comment:
Can we add the survey id and dataset id to the log line, ru_ref/qid should already be bound.

Reply:
Yep - they are already bound, added those two extras 👍